### PR TITLE
[core] generic neighbour elements of elements

### PR DESCRIPTION
**Pull request by pablobecker**
**📝 Description**

New process that finds neighbour elements of elements. 
Note that unlike the original neigh elem of elem, this one returns an nullptr when there's no neigh element in the selected face.

**🆕 Changelog**
- Added generic_find_elements_neighbours_process
-  Added serial test

### CODE REVIEW

**Comment by pablobecker** [comment]
missing MPI test.
I tried reading the mdpa from the cpp test to create an mpi version of the test, but i had issues with the path to the mdpa, so i discarded it

Perhaps I could create a helper function (exposed to python) only for the test, so that I can create a python test and then read correctly an mdpa in mpi

Do you have a better idea?

**Comment by pablobecker** [comment]
ping @KratosMultiphysics/implementation-committee

File: kratos/processes/generic_find_elements_neighbours_process.h
Comment (ddiezrod) [inline kratos/processes/generic_find_elements_neighbours_process.h]: Isnt this done already in InitializeElementNeighbours?

**Comment by ddiezrod** [comment]
> missing MPI test. I tried reading the mdpa from the cpp test to create an mpi version of the test, but i had issues with the path to the mdpa, so i discarded it
> 
> Perhaps I could create a helper function (exposed to python) only for the test, so that I can create a python test and then read correctly an mdpa in mpi
> 
> Do you have a better idea?

But does this work in MPI? From what I see in the code, I dont think so.

**Comment by ddiezrod** [comment]
> > missing MPI test. I tried reading the mdpa from the cpp test to create an mpi version of the test, but i had issues with the path to the mdpa, so i discarded it
> > Perhaps I could create a helper function (exposed to python) only for the test, so that I can create a python test and then read correctly an mdpa in mpi
> > Do you have a better idea?
> 
> But does this work in MPI? From what I see in the code, I dont think so.

I've discussed this with @pablobecker , I had not seen that in the comparison you are only checking the pointer itself, not the elements, so everything should work fine.

File: kratos/processes/generic_find_elements_neighbours_process.h
Comment (AlejandroCornejo) [inline kratos/processes/generic_find_elements_neighbours_process.h]: Do we need this comments?

Comment (pablobecker) [reply kratos/processes/generic_find_elements_neighbours_process.h]: @pooyan-dadvand suggested to keep it for the future. I have no preference here

**Comment by AlejandroCornejo** [comment]
besides we should separate the implementastion in .h and .cpp right?

File: kratos/processes/generic_find_elements_neighbours_process.h
Comment (roigcarlo) [inline kratos/processes/generic_find_elements_neighbours_process.h]: This comparison in particular can be problematic in MPI. If I understand correctly, you are comparing the physical address of the Element to remove duplicates. If you access by `&*` I assume because you want the address of the element to avoid having duplicated elements from different processes, you are calling the overloaded `*` from the global_pointer, which will segfault. use:

```C++
if( PointersOfAllFaceNodes(j).get() == PointersOfAllFaceNodes(k).get() ){
```

This will return the address without accessing the data.

Another solution will be to get the vector with the id's of the pointers via a communicator (like its done in the nodal_neighbor process)

**Comment by miguelmaso** [comment]
> Note that unlike the original neigh elem of elem, this one returns an nullptr when there's no neigh element in the selected face.

IMG_20220512_172349

neighbor elements of 3 with the old process: `[2, 4]`
neighbor elements of 3 with the new process: `[2, nullptr, 4]`
being the first face shared by elems 3-2 and the third face, shared by 3-4

Maybe it is interesting to pass a flag or a bool to the old process in order to select the desired behavior (e.g. `sort_neighbours_by_faces`)

File: kratos/processes/generic_find_elements_neighbours_process.h
Comment (pablobecker) [reply kratos/processes/generic_find_elements_neighbours_process.h]: didnt know, thank you!

**Comment by pablobecker** [comment]
> neighbor elements of 3 with the old process: `[2, 4]`

no, the previous one returned a pointer to the element itself in case of no neigh in the face:
https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/processes/find_elements_neighbours_process.h#L362
so the new and old processes return the neigh elems in the same positions 
 in this case it'd be [2,3,4]

**Comment by RiccardoRossi** [comment]
why returing a ppointer to the element itself is a problem??

**Comment by pablobecker** [comment]
> why returing a ppointer to the element itself is a problem??

IMO it's misleading, the n position of the vector should have the pointer to the element in the n face. returning a pointer to the elem itself forces you to access the element to check if it was a real neighbour or itself

**Comment by RiccardoRossi** [comment]
but then we shall we remove it for good (making the corresponding changes through the code) rather than creating a new object which provides the same capabilities except this detail...

**Comment by miguelmaso** [comment]
> 

> > why returing a ppointer to the element itself is a problem??
> 
> IMO it's misleading, the n position of the vector should have the pointer to the element in the n face. returning a pointer to the elem itself forces you to access the element to check if it was a real neighbour or itself

I agree, it's very misleading. If I had known this, I'd have modified some implementations

> but then we shall we remove it for good (making the corresponding changes through the code) rather than creating a new object which provides the same capabilities except this detail...

and add some doxygen explaining it

**Comment by pablobecker** [comment]
> but then we shall we remove it for good (making the corresponding changes through the code) rather than creating a new object which provides the same capabilities except this detail...

the current one only works with tets, while this one works for all geometries (as long as they have implemented geom.GenerateFaces() )

i can delete the old one, but to keep this PR smaller, i'd like to first merge this one and then in a second PR remove the old one and replace it with this new one, with the necessary changes in all the apps that are using it

**Comment by RiccardoRossi** [comment]
will the numbering remain consistent in the cases in which tets are used?

i am not too positive of the two step approach since it will open the way to no doing the change at all...

**Comment by pablobecker** [comment]
> 


> will the numbering remain consistent in the cases in which tets are used?

yes, they are the same. both the current process and the GenerateFaces() use the same face order

**Comment by rubenzorrilla** [comment]
> will the numbering remain consistent in the cases in which tets are used?
> 
> i am not too positive of the two step approach since it will open the way to no doing the change at all...

I agree about the two steps comment. Note that the addition in this PR has no interaction with the rest of code so there is no point at all in doing it in two steps. Besides this, during some time two utilities aiming at doing the same operation with different behaviour will coexist, something that is in my opinion misleading.

File: kratos/processes/generic_find_elements_neighbours_process.h
Comment (pablobecker) [reply kratos/processes/generic_find_elements_neighbours_process.h]: merged everythin into a single elemental loop

**Comment by pablobecker** [comment]
all fixed! could you please review it now?

**Comment by pablobecker** [comment]
sorry @KratosMultiphysics/dem i mistakenly added you to the reviewrs

File: applications/ShallowWaterApplication/custom_elements/conservative_element.cpp
Comment (miguelmaso) [inline applications/ShallowWaterApplication/custom_elements/conservative_element.cpp:148]: Please, wait a little before merging this branch, let me check if the behavior is unmodified. I will run some validation tests

Comment (pablobecker) [reply applications/ShallowWaterApplication/custom_elements/conservative_element.cpp:148]: ping @miguelmaso

Comment (miguelmaso) [reply applications/ShallowWaterApplication/custom_elements/conservative_element.cpp:148]: sorry for the delay. I've just checked the results of the tests and it is OK

**Review by miguelmaso — APPROVED** [review]
from my side, it is OK

File: kratos/processes/generic_find_elements_neighbours_process.h
Comment (rubenzorrilla) [inline kratos/processes/generic_find_elements_neighbours_process.h:36]: These types must be defined inside the class. Note that you're defining them for the entire Kratos in here!

File: kratos/processes/generic_find_elements_neighbours_process.h
Comment (rubenzorrilla) [inline kratos/processes/generic_find_elements_neighbours_process.h:53]: Documentation is missing. I think it is mandatory for such kind of processes.

File: kratos/processes/generic_find_elements_neighbours_process.h
Comment (rubenzorrilla) [inline kratos/processes/generic_find_elements_neighbours_process.h:139]: `mrModelPart` according to style guide.

File: kratos/processes/generic_find_elements_neighbours_process.h
Comment (rubenzorrilla) [inline kratos/processes/generic_find_elements_neighbours_process.h:151]: Why protected if there is no derived class?

File: kratos/processes/generic_find_elements_neighbours_process.h
Comment (rubenzorrilla) [inline kratos/processes/generic_find_elements_neighbours_process.h:192]: This must be in a `private` section. Note that it is now protected as the `private` keyword is missing.

File: kratos/processes/generic_find_elements_neighbours_process.h
Comment (pablobecker) [reply kratos/processes/generic_find_elements_neighbours_process.h:36]: ok, fixing it

File: kratos/processes/generic_find_elements_neighbours_process.h
Comment (pablobecker) [reply kratos/processes/generic_find_elements_neighbours_process.h:53]: ok, adding it

File: kratos/processes/generic_find_elements_neighbours_process.h
Comment (pablobecker) [reply kratos/processes/generic_find_elements_neighbours_process.h:139]: ok

File: kratos/processes/generic_find_elements_neighbours_process.h
Comment (pablobecker) [reply kratos/processes/generic_find_elements_neighbours_process.h:151]: ok

File: kratos/processes/generic_find_elements_neighbours_process.h
Comment (pablobecker) [reply kratos/processes/generic_find_elements_neighbours_process.h:192]: ok

File: kratos/processes/generic_find_elements_neighbours_process.cpp
Comment (rubenzorrilla) [inline kratos/processes/generic_find_elements_neighbours_process.cpp:34]: `auto` type is advisable to be used in here. Besides, variable should be named as `r_geom` according to our style guide.

File: kratos/processes/generic_find_elements_neighbours_process.cpp
Comment (rubenzorrilla) [inline kratos/processes/generic_find_elements_neighbours_process.cpp:35]: `const auto elem_boundaries = ....`

Comment (pablobecker) [reply kratos/processes/generic_find_elements_neighbours_process.cpp:35]: ok

File: kratos/processes/generic_find_elements_neighbours_process.cpp
Comment (pablobecker) [reply kratos/processes/generic_find_elements_neighbours_process.cpp:34]: ok

File: kratos/processes/generic_find_elements_neighbours_process.cpp
Comment (rubenzorrilla) [inline kratos/processes/generic_find_elements_neighbours_process.cpp:42]: Considering that you're calling the `resize` below, I think that this can be substituted by a call to `clear`.

File: kratos/processes/generic_find_elements_neighbours_process.cpp
Comment (rubenzorrilla) [inline kratos/processes/generic_find_elements_neighbours_process.cpp:50]: Wouldn't be sufficient to do the `reserve` in here?

File: kratos/processes/generic_find_elements_neighbours_process.cpp
Comment (rubenzorrilla) [inline kratos/processes/generic_find_elements_neighbours_process.cpp:62]: Formatting issues.

File: kratos/processes/generic_find_elements_neighbours_process.cpp
Comment (rubenzorrilla) [inline kratos/processes/generic_find_elements_neighbours_process.cpp:70]: `auto` type is preferred in here.

File: kratos/processes/generic_find_elements_neighbours_process.cpp
Comment (rubenzorrilla) [inline kratos/processes/generic_find_elements_neighbours_process.cpp:134]: Please remove commented code.

File: kratos/processes/generic_find_elements_neighbours_process.cpp
Comment (rubenzorrilla) [inline kratos/processes/generic_find_elements_neighbours_process.cpp:141]: `auto`

**Review by rubenzorrilla — COMMENTED** [review]
There are still some minors. Sorry, but IMO we should be a bit picky with key utilities like this one. 
Besides, there are several style guide and formatting issues.

Thanks for the patience.

File: kratos/processes/generic_find_elements_neighbours_process.cpp
Comment (pablobecker) [reply kratos/processes/generic_find_elements_neighbours_process.cpp:42]: ok

File: kratos/processes/generic_find_elements_neighbours_process.cpp
Comment (pablobecker) [reply kratos/processes/generic_find_elements_neighbours_process.cpp:50]: no, i am not doing push_back later and size must always be nBoundaries

File: kratos/processes/generic_find_elements_neighbours_process.cpp
Comment (maceligueta) [reply kratos/processes/generic_find_elements_neighbours_process.cpp:34]: @rubenzorrilla  what is the rule of thumb on using `auto` or standard declaration?

**Comment by pablobecker** [comment]
thanks for the review. Unfortunately i didnt expect the TC to jump in and merged after i got the approval. Apologies, i was not aware of the protocol for PRs like these.
I created anew  branch with the fixed and will push it once it's compiled locally and tests pass

File: kratos/processes/generic_find_elements_neighbours_process.cpp
Comment (pablobecker) [reply kratos/processes/generic_find_elements_neighbours_process.cpp:134]: ok

File: kratos/processes/generic_find_elements_neighbours_process.cpp
Comment (pablobecker) [reply kratos/processes/generic_find_elements_neighbours_process.cpp:141]: ok

File: kratos/processes/generic_find_elements_neighbours_process.cpp
Comment (rubenzorrilla) [reply kratos/processes/generic_find_elements_neighbours_process.cpp:34]: At the end there is no rule at all, but it is advisable for complex types, like the one in here, in which the `auto` becomes very handy. Imagine the situation in which we finally get rid of the template argument of the `Node`, as we did in the  past with the `Point`, we'll need to  go one by one changing the declarations. If `auto` is used, this becomes automatic.

On the contrary, we tend to avoid it when using simple types like `double`, `int` and similars. Besides, I've experienced some weird missbehaviors when doing algebraic operations with variables, so I also tend to avoid it with `array_1d<double,3>`-like types.

As a rule of thumb, I'd say this.

**Comment by rubenzorrilla** [comment]
> thanks for the review. Unfortunately i didnt expect the TC to jump in and merged after i got the approval. Apologies, i was not aware of the protocol for PRs like these. I created anew branch following your review and will push it once it's compiled locally and tests pass

OK. Also note that `test_generic_neighbour_elements.cpp` are failing in the nightly build. I think this is potentially related to this.

**Comment by rubenzorrilla** [comment]
I'm also getting a linking error when compiling in Windows...

**Comment by rubenzorrilla** [comment]
> I'm also getting a linking error when compiling in Windows...

Strange, it didn't happen anymore...

**Pull request merged**
